src/ModelContextProtocol.Core/Client/StdioClientTransport.cs

```
@@ -199,7 +200,13 @@ public async Task<ITransport> ConnectAsync(CancellationToken cancellationToken =
             throw new IOException("Failed to connect transport.", ex);
         }
     }
+    [DllImport("kernel32.dll")]
+    private static extern IntPtr GetConsoleWindow();
 
+    private static bool HasConsole()
+    {
+        try { return GetConsoleWindow() != IntPtr.Zero; } catch { return false; }
```

I'm very suspicious this is the best way to determine whether the Console.InputEncoding setter is going to throw a NotSupportedException.

If we're going to catch arbitrary exceptions anyway, it would be better to catch the NotSupportedException from the setter although I'm not sure that's the best approach either. I think #73 added this logic. I'm curious if @willibrandon  or @stephentoub have alternatives.

---

> If we're going to catch arbitrary exceptions anyway, it would be better to catch the NotSupportedException from the setter

Yeah, I'd rather just catch/eat the exception from the actual operation being performed than have a separate IsXx method that does the exact same thing and that's then used as a guard. We're about to launch a whole process; I'm not particularly concerned at the overhead of an exception.

That said, I'm unclear about the originally stated problem. I just created a WPF app, set both Console.OutputEncoding and Console.InputEncoding, and it worked fine, no exceptions. The code for these properties in Console also just bails without exception for what I would have expected in these cases:
https://github.com/dotnet/runtime/blob/edb570cd9e75036b6c61b201a0a9a9f2301db276/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L135-L138

Can you share a repro of the issue?